### PR TITLE
Added warning for unused arithmetic expressions

### LINKED ISSUE #50124: Warning for unused arithmetic expressions
After having spent a day tracking down [this bug](https://github.com/mit-pdos/distributary/blob/c7a922e082b7862c94a8ef06b25048ecfc237420/dataflow/src/backlog/multiw.rs#L116), and recalling all the past times the same thing has bitten me, I'd like to propose a lint unused results of arithmetic expressions. Consider the following code:
```rust
let mut delta = 0isize;
match foo {
    Foo::Add(x) => {
        delta += x;
    }
    Foo::Sub(x) => {
        delta - x;
    }
}
```
This code currently compiles with no warnings (given a suitable `enum Foo`). However, bugs like these can be extremely hard to track down. They can often be hard to spot, especially when going through methods like `saturating_sub` which do not *modify* `self`, but instead return the new result (e.g., `delta = delta.saturating_sub(x)` is necessary).

It would be really neat to have a warning about unused arithmetic expressions (ideally one that also handles `saturating_sub`). I don't think there are generally cases where not using the result of an arithmetic expression is what the developer intended to do. This also complements the `unused_variables` and `unused_mut` lints quite nicely.

### PR DESCRIPTION
**Pull request by aaronaaeng**
The compiler now displays a warning when a binary arithmetic operation is evaluated but not used.  This resolves #50124  by following the instructions outlined in the issue.  The changes are as follows:

- Added new pattern matching for unused arithmetic expressions in `src/librustc_lint/unused.rs`
- Added `#[must_use]` attributes to the binary operation methods in `src/libcore/internal_macros.rs`
- Added `#[must_use]` attributes to the non-assigning binary operators in `src/libcore/ops/arith.rs`

### CODE REVIEW

File: src/libcore/internal_macros.rs [hunk 1]
```diff
@@ -38,6 +38,7 @@ macro_rules! forward_ref_binop {
     };
     (impl $imp:ident, $method:ident for $t:ty, $u:ty, #[$attr:meta]) => {
         #[$attr]
+        #[must_use]
```
Comment (scottmcm) [inline src/libcore/internal_macros.rs]: Is the `impl` the right place for this attribute?  I thought the lint only looked at methods on trait definitions.

File: src/librustc_lint/unused.rs [hunk 2]
```diff
@@ -103,6 +103,11 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
                         op_warned = true;
                     },
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
+                        let msg = "unused arithmetic operation which must be used";
```
Comment (scottmcm) [inline src/librustc_lint/unused.rs]: 1. While you're here, maybe add the bitops too?  `a^b;` giving a warning also makes sense to me
2. Unary ops too? `-x` seems to also fit "unused arithmetic operation which must be used"
3. At that point, is it just all ops?  The only three left are `*`, `&&`, and `||`.  (The short-circuiting ones I could see an argument for not having the lint , since `foo() || panic!()` is possible, but...)

File: src/libcore/internal_macros.rs (hunk 1 above)
Comment (varkor) [reply src/libcore/internal_macros.rs]: Sorry, I didn't read the mentoring instructions before, but this fix needs tweaking:
- Instead of adding `#[must_use]` to method definitions in impls, you need to add them to the trait definitions (e.g. here: https://github.com/rust-lang/rust/blob/699eb95ae3dae023cddca85facf6660b196bd0d1/src/libcore/ops/arith.rs#L97).
- You don't need any of the `#[must_use]` in the macros, for the same reason.

Comment (aaronaaeng) [reply src/libcore/internal_macros.rs]: I don't fully understand what is going on in the `impl`, though I'm a little unsure of where else I would put the attribute.  The recurrence of `#[$attr]` in each `impl` made it seem as though the attributes need to be repeated before each possible combination of `T` and `U` states.

Comment (aaronaaeng) [reply src/libcore/internal_macros.rs]: Oh, ok.  I can go fix that.  Thanks for the quick support.  I also just learned that even though GitHub will automatically show new messages as they come in, I need to refresh to see edits.

Comment (varkor) [reply src/libcore/internal_macros.rs]: Take a look at `PartialOrd`, which implements `lt` (`<`): https://github.com/rust-lang/rust/blob/699eb95ae3dae023cddca85facf6660b196bd0d1/src/libcore/cmp.rs#L642-L650
The `#[must_use]` attribute is on the definition in `trait PartialOrd`, but not on any of the `impl PartialOrd`. The way `#[must_use]` works at the moment, it's only the trait declaration which is important — everything else inherits from the trait. (That way, _anyone_ who implements an operator will get the advantage of `#[must_use]` without having to add it themselves.

File: src/librustc_lint/unused.rs [hunk 3]
```diff
@@ -103,6 +103,11 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
                         op_warned = true;
                     },
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
```
[the remark this answers is not in the document]
Comment (varkor) [reply src/librustc_lint/unused.rs]: Yeah, different messages sound good 👍
We want the remaining operators too, though (the same message as the other non-comparison operators is fine).

File: src/librustc_lint/unused.rs (hunk 2 above)
Comment (aaronaaeng) [reply src/librustc_lint/unused.rs]: I'm working on the `UnaryOp` pattern matching.  While matching on binary operators, it matches on `bin_op.node`.  However, the `UnOp` type doesn't have this field.  I've tried to find the definition of this type but I've come up with nothing so far.  Any suggestion as to what file I should check out?

Comment (abonander) [reply src/librustc_lint/unused.rs]: `UnOp` is an enum [defined here](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/mod.rs#L1054). `BinOp` is an enum as well but it's [wrapped in a struct that gives it span information](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/mod.rs#L1054), which I assume was considered unnecessary for unary ops.

All the `hir::` types are defined in `src/librust/hir.mod.rs`. For convenience all the enums have their variants reexported in this module as well.

**Comment by aaronaaeng** [comment]
I made the changes discussed.  Warnings are now thrown whenever an operator is evaluated but not used.  The warning is specific to the type of operator that went unused.  The `#[must_use]` attribute has been moved to the correct location for each operator.  It passes all of the `x.py` tests and works as intended with sample code that evaluates expressions using binary and unary operators but does not use the results.

**Comment by abonander** [comment]
Putting `#[must_use]` on the traits themselves is probably the right move but I don't know what the policy is for expanding the scope of lints. If someone has `#[deny(unused_must_use)]` in their crate and they made a mistake with one of these operators, this change will break their build. It's a *good* breakage, but a breakage nonetheless.

File: src/librustc_lint/unused.rs [hunk 4]
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                        op_warned = true;
+                    },
+                    hir::BiBitXor | hir::BiBitAnd | hir::BiBitOr | hir::BiShl | hir::BiShr => {
+                        let msg = "unused bitwise operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
+                    },
+                }
+            }
+            if let hir::ExprUnary(un_op, ..) = expr.node {
```
Comment (abonander) [inline src/librustc_lint/unused.rs]: I would combine this and the `if let hir::ExprBinary` bit above into a single `match`.

File: src/librustc_lint/unused.rs [hunk 5]
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
                         op_warned = true;
                     },
-                    _ => {},
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
+                        let msg = "unused arithmetic operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
```
Comment (abonander) [inline src/librustc_lint/unused.rs]: I would extract the above two lines out of each arm and just have each arm return its relevant message and then emit the lint at the end.

File: src/librustc_lint/unused.rs [hunk 6]
```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                    },
+                    hir::BiAnd | hir::BiOr => {
+                        let msg = "unused logical operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
```
Comment (abonander) [inline src/librustc_lint/unused.rs]: For bonus points you could emit a help message here saying to use `if (first_op) { second_op }` instead, but that's a bit involved. I might tack that on myself if you don't want to get into it.

**Comment by abonander** [comment]
I don't see a lot of UI tests for lints in general so I don't know how we want to test these additions.

File: src/librustc_lint/unused.rs (hunk 4 above)
Comment (aaronaaeng) [reply src/librustc_lint/unused.rs]: What is the best practice for combining these two match statements?  Combining them seems like it would require nested matches since `ExprBinary` and `ExprUnary` are different types.  I have to match on `op.node` for binary operators but only on `op` for unary operators.  Should I implement this nested system or is there a better way?

Comment (abonander) [reply src/librustc_lint/unused.rs]: A nested match is fine. It's mostly about eliminating the redundant code here.

**Comment by zackmdavis** [comment]
> Putting `#[must_use]` on the traits themselves is probably the right move but

Note that `#[must_use]` for functions is still unstable. (Although probably not for long: the stabilization final comment period [ended](https://github.com/rust-lang/rust/issues/43302#issuecomment-380311079) with no objections, and who can say but that I might even have time to rebase the [stabilization PR](https://github.com/rust-lang/rust/pull/48925) tomorrow??)

> I don't know what the policy is for expanding the scope of lints

Given that [RFC 1193](https://github.com/rust-lang/rfcs/blob/master/text/1193-cap-lints.md) lint capping exists specifically to preven the breakage from spreading to dependent crates, it's probably OK sometimes (but is definitely OK here given `fn_must_use` not even being stable). (See #48717 for an example of Crater-detected breakage from a lint behavior change.)

> I don't see a lot of UI tests for lints in general

There's a src/test/ui/lint directory, but this would probably go with rust/src/test/ui/rfc_1940-must_use_on_functions/fn_must_use.rs (which should be moved to src/test/ui/lint/ on stabilization).

**Comment by abonander** [comment]
@zackmdavis This change also affects the arithmetic, the logical/bitwise operators and the unary operators. Where would those lint tests go.

**Comment by zackmdavis** [comment]
new file `src/test/ui/lint/must_use_ops.rs`??

**Comment by aaronaaeng** [comment]
I updated the match statements to reduce redundant code.  Because I am not incredibly familiar with pattern matching, I introduced a bool that says whether to send the lint message.  Without this check, it could throw empty warnings on expressions that were correct.

File: src/librustc_lint/unused.rs [hunk 7]
```diff
@@ -91,23 +93,39 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
-                    // the comparison trait methods
-                    hir::BiEq | hir::BiLt | hir::BiLe | hir::BiNe | hir::BiGe | hir::BiGt => {
-                        let msg = "unused comparison which must be used";
-                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
-                        op_warned = true;
-                    },
-                    _ => {},
+            match expr.node {
+                hir::ExprBinary(bin_op, ..)  => {
+                    should_warn_op = true;
```
Comment (zackmdavis) [inline src/librustc_lint/unused.rs]:
```rust
let must_use_op = match expr.node {
    hir::ExprBinary(bin_op, ..)  => {
        match bin_op.node {
            hir::BiEq | hir::BiLt | hir::BiLe | hir::BiNe | hir::BiGe | hir::BiGt => {
                Some("comparison")
            },
            hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
                Some("arithmetic operation")
            },
            hir::BiAnd | hir::BiOr => {
                Some("logical operation")
            },
            hir::BiBitXor | hir::BiBitAnd | hir::BiBitOr | hir::BiShl | hir::BiShr => {
                Some("bitwise operation")
            },
        }
    },
    hir::ExprUnary(..) => {
        Some("unary operation")
    },
    _ => None
}
if let Some(must_use_op) {
    cx.span_lint(UNUSED_MUST_USE, expr.span, format!("unused {} which must be used", must_use_op));
    op_warned = true;
}
```

**Comment by aaronaaeng** [comment]
After seeing that, it seemed pretty obvious.  I made the necessary changes and it is a lot cleaner than it was.  Hopefully it all looks good now.

File: src/librustc_lint/unused.rs [hunk 8]
```diff
@@ -91,23 +91,34 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                        },
+                        hir::BiAnd | hir::BiOr => {
+                            Some("logical operation")
+                        },
+                        hir::BiBitXor | hir::BiBitAnd | hir::BiBitOr | hir::BiShl | hir::BiShr => {
+                            Some("bitwise operation")
+                        },
+                    }
+                },
+                hir::ExprUnary(..) => {
```
Comment (zackmdavis) [inline src/librustc_lint/unused.rs]: I just realized we can save two lines by writing this as `hir::ExprUnary(..) => Some("unary operation"),` rather than with a block (obviously this is a criticism of my earlier review comment, not you)

**Comment by zackmdavis** [comment]
Needs [tests](https://rust-lang-nursery.github.io/rustc-guide/tests/adding.html) (probably a new UI test file at src/test/ui/lint/must_use_ops.rs or similar; don't forget the `#[feature(fn_must_use)]`), 

For the benefit of future software-archeologists, I argue that it's also worth [squashing](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) these 6+ commits down into 1. (GitHub truncates the commit message summary line at 72 characters, so something like "make must-use lint cover binary operations" (42 chars) is better than 12d227aff's 95 chars.)

**Comment by aaronaaeng** [comment]
Should I include each operator or just one from each category, i.e., do I need to include an expected warning for `+`, `-`, `*`, and `/`, or should I just include one operator from the arithmetic category?

**Comment by aaronaaeng** [comment]
My attempt at a rebase did not fix the message as hoped.  Let me try again.

**Comment by zackmdavis** [comment]
looks good to me (but I don't have the authority to approve PRs)

**Comment by zackmdavis** [comment]
>  wants to merge 1 commit into rust-lang:master from aaronaaeng:master

(Standard practice is to start a new branch in your own repository, rather than working on master; this surprised me just now as I began to rebase #48925 off this.)

**Comment by aaronaaeng** [comment]
That makes sense.  I'll do that next time I submit a PR somewhere.  Thank you for helping me out with all of this.

**Comment by leonardo-m** [comment]
In the last Nightly I still don't see those warnings. Perhaps I am missing something?

**Comment by kennytm** [comment]
@leonardo-m You need `#![feature(fn_must_use)]` to see those warnings for now. After #48925 is merged (within 1 to 2 days) they should be enabled automatically.
